app/models/lead.rb

```diff
@@ -0,0 +1,144 @@
+# Leads models bifurcating keys; each row represents a single couplet option in a key.
+#
+# @!attribute parent_id
+#   @return [integer]
+#   Id of the lead immediately prior to this one in the key
+#
+# @!attribute otu_id
+#   @return [integer]
+#   Otu determined at this stage of the key, if any
+#
+# @!attribute text
+#   @return [text]
+#   Text for this option of a couplet in the key
+#
+# @!attribute origin_label
+#   @return [string]
+#   If the couplet was given a # in print, that number. If provided assume there is an origin citation.
+#
+# @!attribute description
+#   @return [text]
+#   Only used on the starting couplet, to describe the overall key
+#
+# @!attribute redirect_id
+#   @return [integer]
+#   Id of the lead to redirect to if this option of the couplet is chosen
+#
+# @!attribute link_out
+#   @return [text]
+#   Provides an alternate URL to display as a terminal link
+#
+# @!attribute link_out_text
+#   @return [string]
+#   Text to display for the link_out URL
+#
+# @!attribute position
+#   @return [integer]
+#   A sort order used by acts_as_list and has_closure_tree
+#
+# @!attribute is_public
+#   @return [boolean]
+#   True if the key is viewable without being logged in
+#
+class Lead < ApplicationRecord
+  include Housekeeping
+  include Shared::Tags
+  include Shared::Citations
+  include Shared::IsData
+
+  belongs_to :parent, class_name: 'Lead'
+  acts_as_list scope: :parent
+  belongs_to :otu, inverse_of: :leads
+  belongs_to :redirect, class_name: 'Lead'
+
+  has_closure_tree order: 'position', numeric_order: true, dont_order_roots: true
+
+  def display_name
+    id.to_s + ': ' + text.slice(0..40) + (text.size > 40 ? '...' : '')
+  end
+
+  def future
+    self.redirect_id.blank? ? self.all_children : self.redirect.all_children
+  end
+
+  def go_id
+    (self.redirect_id.presence || self.id)
+  end
+
+  def dupe(node = self, id = nil)
+    a = node.dup
+    a.parent_id = id
+    a.description = (a.description + ' (COPY)') if id == nil
+    a.save!
```

Comment:
If we continue with this method we should wrap this in a transaction and rescue ActiveRecord::RecordInvalid (all `!` can raise).